File: convert/ollaya_convert/arch.py
"""The `arch` layer (`application/vnd.ollaya.arch`): what the MLX engine builds its network from.

    python3 -m ollaya_convert.arch laya --config encoder/config.json --agent-config rl_agent_config.json \\
        --weights model.safetensors --out arch.json
    python3 -m ollaya_convert.arch sequence-classification --config config.json --weights model.safetensors --out arch.json
    python3 -m ollaya_convert.arch von --config config.json --weights option_marker.pt --out arch.json
    python3 -m ollaya_convert.arch gliclass --config config.json --weights model.safetensors --out arch.json

The MLX engine (`crates/ollaya-runner/src/mlx/`) runs our own code per backbone and head on the
author's unmodified weights file. This layer tells it which network that file holds: the encoder's
hyperparameters from the author's `config.json`, the head's from the author's model code (named
below), and the tensor names. It holds no weights. A model without it runs on ONNX Runtime only.

Every tensor the Rust loader reads is checked against the weights file here, so a wrong prefix
fails at build time. PyTorch zip checkpoints (von) also get each tensor's byte offset: every
storage is an uncompressed zip member, and the pickle is read here with a restricted unpickler
that resolves only tensor-rebuild records (as `families/weightless_ext.py` does), so the runtime
never reads a pickle.

Standard library only: this runs anywhere, without the conversion environment.
"""
import argparse
import collections
import json
import pickle
import struct
import zipfile

SCHEMA = 1
DTYPE_SIZES = {"F32": 4, "F16": 2, "BF16": 2, "I64": 8}
TORCH_STORAGES = {"FloatStorage": "F32", "HalfStorage": "F16", "BFloat16Storage": "BF16", "LongStorage": "I64"}


# ------------------------------------------------------------------------------ weights indexes


def safetensors_index(path):
    """{name: (dtype, shape)} from a safetensors header."""
    with open(path, "rb") as f:
        n = struct.unpack("<Q", f.read(8))[0]
        header = json.loads(f.read(n))
    header.pop("__metadata__", None)
    return {k: (v["dtype"], tuple(v["shape"])) for k, v in header.items()}


_Storage = collections.namedtuple("_Storage", "dtype key numel")
_Tensor = collections.namedtuple("_Tensor", "storage offset size stride")


def _rebuild_tensor_v2(storage, storage_offset, size, stride, requires_grad=False, backward_hooks=None, metadata=None):
    return _Tensor(storage, storage_offset, tuple(size), tuple(stride))


class _MetadataUnpickler(pickle.Unpickler):
    """Resolves only what a state_dict pickle needs; any other global is refused."""

    def find_class(self, module, name):
        if (module, name) == ("torch._utils", "_rebuild_tensor_v2"):
            return _rebuild_tensor_v2
        if (module, name) == ("collections", "OrderedDict"):
            return collections.OrderedDict
        if module == "torch" and name in TORCH_STORAGES:
            return name
        raise pickle.UnpicklingError("refusing to resolve %s.%s" % (module, name))

    def persistent_load(self, pid):
        kind, storage_type, key, _location, numel = pid
        if kind != "storage" or storage_type not in TORCH_STORAGES:
            raise pickle.UnpicklingError("unsupported storage %r" % (pid,))
        return _Storage(TORCH_STORAGES[storage_type], key, numel)


def torch_zip_index(path):
    """{name: (dtype, shape, absolute offset)} for every contiguous tensor of a torch.save zip."""
    zf = zipfile.ZipFile(path)
    pkl = next(n for n in zf.namelist() if n.endswith("/data.pkl"))
    root = pkl[: -len("data.pkl")]
    obj = _MetadataUnpickler(zf.open(pkl)).load()
    starts = {}
    with open(path, "rb") as f:
        for info in zf.infolist():
            if not info.filename.startswith(root + "data/"):
                continue
            if info.compress_type != zipfile.ZIP_STORED:
                raise SystemExit("%s is compressed; cannot reference it" % info.filename)
            f.seek(info.header_offset)
            hdr = f.read(30)
            n_name, n_extra = struct.unpack("<HH", hdr[26:30])
            starts[info.filename[len(root) + 5:]] = info.header_offset + 30 + n_name + n_extra
    index = {}
    for name, t in obj.items():
        if not isinstance(t, _Tensor):
            continue
        expect, acc = [], 1
        for d in reversed(t.size):
            expect.append(acc)
            acc *= d
        if tuple(reversed(expect)) != tuple(t.stride):
            continue  # not contiguous: cannot be read in place
        item = DTYPE_SIZES[t.storage.dtype]
        index[name] = (t.storage.dtype, t.size, starts[t.storage.key] + t.offset * item)
    return index


# ------------------------------------------------------------------------------------- backbone


def modernbert(config, prefix):
    """Hugging Face ModernBertConfig, normalised to transformers 5 (layer types, RoPE per type).

    Older configs (transformers 4, e.g. Ettin in GLiClass) give `global_attn_every_n_layers`,
    `global_rope_theta` and `local_rope_theta`; transformers 5 derives the same values from them
    (configuration_modernbert.py), with defaults 3, 160000 and 10000."""
    c = config
    n = c["num_hidden_layers"]
    types = c.get("layer_types") or [
        "sliding_attention" if i % c.get("global_attn_every_n_layers", 3) else "full_attention" for i in range(n)]
    rope = c.get("rope_parameters") or {}
    theta = {
        "full_attention": float((rope.get("full_attention") or {}).get("rope_theta", c.get("global_rope_theta", 160000.0))),
        "sliding_attention": float((rope.get("sliding_attention") or {}).get("rope_theta", c.get("local_rope_theta", 10000.0))),
    }
    for kind, params in rope.items():
        if (params or {}).get("rope_type", "default") != "default":
            raise SystemExit("unsupported rope_type for %s: %r" % (kind, params))
    if c.get("model_type") != "modernbert":
        raise SystemExit("not a ModernBERT config: %r" % c.get("model_type"))
    return {
        "type": "modernbert",
        "prefix": prefix,
        "vocab_size": c["vocab_size"],
        "hidden_size": c["hidden_size"],
        "intermediate_size": c["intermediate_size"],
        "num_hidden_layers": n,
        "num_attention_heads": c["num_attention_heads"],
        "norm_eps": c.get("norm_eps", 1e-5),
        "norm_bias": c.get("norm_bias", False),
        "attention_bias": c.get("attention_bias", False),
        "mlp_bias": c.get("mlp_bias", False),
        "hidden_activation": c.get("hidden_activation", "gelu"),
        "local_attention": c.get("local_attention", 128),
        "layer_types": types,
        "rope_theta": theta,
    }


def modernbert_tensors(b):
    """Every tensor `crates/ollaya-runner/src/mlx/modernbert.rs` reads."""
    p = b["prefix"]
    names = [p + "embeddings.tok_embeddings.weight", p + "embeddings.norm.weight", p + "final_norm.weight"]
    for i in range(b["num_hidden_layers"]):
        lp = "%slayers.%d." % (p, i)
        if i:
            names.append(lp + "attn_norm.weight")
        names += [lp + "attn.Wqkv.weight", lp + "attn.Wo.weight", lp + "mlp_norm.weight",
                  lp + "mlp.Wi.weight", lp + "mlp.Wo.weight"]
    return names


# ----------------------------------------------------------------------------------------- heads


def laya(config, agent):
    """laya 0.3.7 `laya.common.DecisionModel`: `nn.TransformerEncoderLayer(d, max(1, d // 64),
    4 * d, dropout, batch_first=True, norm_first=True)` x `head_layers` (PyTorch defaults: ReLU,
    eps 1e-5), `type_emb`, `scorer` (LayerNorm, Linear, GELU, Linear) and `act_head`."""
    d = config["hidden_size"]
    head = {
        "type": "laya",
        "type_embeddings": "type_emb.weight",
        "layers_prefix": "head.layers.",
        "layers": agent.get("head_layers", 2),
        "num_heads": max(1, d // 64),
        "activation": "relu",
        "norm_first": True,
        "layer_norm_eps": 1e-5,
        "scorer_prefix": "scorer.",
        "scorer_norm_eps": 1e-5,
        "act_prefix": "act_head.",
    }
    names = [head["type_embeddings"], "scorer.0.weight", "scorer.0.bias", "scorer.1.weight", "scorer.1.bias",
             "scorer.3.weight", "scorer.3.bias", "act_head.0.weight", "act_head.0.bias", "act_head.2.weight",
             "act_head.2.bias"]
    for i in range(head["layers"]):
        lp = "head.layers.%d." % i
        names += [lp + n for n in ("self_attn.in_proj_weight", "self_attn.in_proj_bias", "self_attn.out_proj.weight",
                                   "self_attn.out_proj.bias", "linear1.weight", "linear1.bias", "linear2.weight",
                                   "linear2.bias", "norm1.weight", "norm1.bias", "norm2.weight", "norm2.bias")]
    return modernbert(config, "encoder."), head, names


def sequence_classification(config):
    """transformers `ModernBertForSequenceClassification`: pooling, `head` (dense, activation,
    norm), `classifier`."""
    head = {
        "type": "sequence-classification",
        "head_prefix": "head.",
        "classifier_prefix": "classifier.",
        "pooling": config.get("classifier_pooling", "cls"),
        "activation": config.get("classifier_activation", "gelu"),
        "classifier_bias": config.get("classifier_bias", False),
        "num_labels": len(config["id2label"]),
    }
    names = ["head.dense.weight", "head.norm.weight", "classifier.weight", "classifier.bias"]
    if head["classifier_bias"]:
        names.append("head.dense.bias")
    return modernbert(config, "model."), head, names


def von(config):
    """von-sdk 1.1.1 `von.models.option_marker.OptionMarkerScorer`: LayerNorm (eps 1e-5), Linear
    to hidden / 2, GELU, LayerNorm, Linear to 1, at every option marker."""
    head = {"type": "option-marker", "prefix": "scorer.", "norm_eps": 1e-5}
    names = ["scorer." + n for n in ("input_norm.weight", "input_norm.bias", "dense.weight", "dense.bias",
                                     "norm.weight", "norm.bias", "out_proj.weight", "out_proj.bias")]
    return modernbert(config, "encoder."), head, names


def gliclass(config):
    """gliclass 0.1.20 `GLiClassUniEncoder` (uni-encoder, prompt first): segment embeddings added to
    the word embeddings, label-span mean pooling (`class_token_pooling` average), `FeaturesProjector`
    x 2, `MLPScorer`."""
    if config.get("architecture_type") != "uni-encoder" or not config.get("use_segment_embeddings"):
        raise SystemExit("only the uni-encoder with segment embeddings is supported")
    if config.get("normalize_features") or config.get("use_lstm") or config.get("extract_text_features"):
        raise SystemExit("unsupported GLiClass options")
    if config.get("class_token_pooling") != "average":
        raise SystemExit("unsupported class_token_pooling %r" % config.get("class_token_pooling"))
    head = {
        "type": "gliclass-uni",
        "text_token_id": config["text_token_index"],
        "segment_embeddings": "model.segment_embeddings.weight",
        "pooling": config.get("pooling_strategy", "first"),
        "text_projector": "model.text_projector.",
        "classes_projector": "model.classes_projector.",
        "projector_activation": config.get("projector_hidden_act", "gelu"),
        "scorer_type": config.get("scorer_type", "mlp"),
        "scorer_prefix": "model.scorer.",
    }
    names = [head["segment_embeddings"]]
    for p in ("model.text_projector.", "model.classes_projector."):
        names += [p + n for n in ("linear_1.weight", "linear_1.bias", "linear_2.weight", "linear_2.bias")]
    names += ["model.scorer.mlp.%d.%s" % (i, n) for i in (0, 2, 4) for n in ("weight", "bias")]
    encoder = dict(config["encoder_config"], vocab_size=config.get("vocab_size", config["encoder_config"]["vocab_size"]))
    return modernbert(encoder, "model.encoder_model."), head, names


# ----------------------------------------------------------------------------------------- build


def build(family, config, weights_path, agent_config=None):
    if family == "laya":
        backbone, head, names = laya(config, agent_config or {})
    elif family == "sequence-classification":
        backbone, head, names = sequence_classification(config)
    elif family == "von":
        backbone, head, names = von(config)
    elif family == "gliclass":
        backbone, head, names = gliclass(config)
    else:
        raise SystemExit("unknown family %r" % family)
    names += modernbert_tensors(backbone)

    torch_zip = zipfile.is_zipfile(weights_path)
    index = torch_zip_index(weights_path) if torch_zip else safetensors_index(weights_path)
    missing = [n for n in names if n not in index]
    if missing:
        raise SystemExit("%s lacks %d tensors, e.g. %s" % (weights_path, len(missing), missing[:3]))
    bad = [n for n in names if index[n][0] not in ("F32", "F16", "BF16")]
    if bad:
        raise SystemExit("unsupported dtypes: %s" % bad[:3])
    b = backbone
    emb = index[b["prefix"] + "embeddings.tok_embeddings.weight"][1]
    if tuple(emb) != (b["vocab_size"], b["hidden_size"]):
        raise SystemExit("embedding shape %s does not match the config" % (emb,))

    weights = {"format": "torch-zip" if torch_zip else "safetensors",
               "file": weights_path.replace("\\", "/").rsplit("/", 1)[-1]}
    if torch_zip:
        weights["tensors"] = {n: {"dtype": index[n][0], "shape": list(index[n][1]), "offset": index[n][2]}
                              for n in sorted(names)}
    return {"schema": SCHEMA, "backbone": backbone, "head": head, "weights": weights}


def main():
    ap = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("family", choices=["laya", "sequence-classification", "von", "gliclass"])
    ap.add_argument("--config", required=True, help="the author's config.json")
    ap.add_argument("--agent-config", help="laya: rl_agent_config.json")
    ap.add_argument("--weights", required=True, help="the author's weights file (safetensors or torch zip)")
    ap.add_argument("--out", required=True)
    a = ap.parse_args()
    config = json.load(open(a.config))
    agent = json.load(open(a.agent_config)) if a.agent_config else None
    arch = build(a.family, config, a.weights, agent)
    with open(a.out, "w") as f:
        json.dump(arch, f, indent=1)
        f.write("\n")
    print("wrote %s: %s + %s head, %s weights, %d tensor offsets" % (
        a.out, arch["backbone"]["type"], arch["head"]["type"], arch["weights"]["format"],
        len(arch["weights"].get("tensors", {}))))


if __name__ == "__main__":
    main()

File: crates/ollaya-runner/src/mlx/arch.rs
//! The `arch` layer (`application/vnd.ollaya.arch`): what the MLX engine needs to build a network
//! from the author's unmodified weights file.
//!
//! It is small, derived at build time (`convert/ollaya_convert/arch.py`) from the author's
//! `config.json` and model code, and holds only hyperparameters and tensor names:
//!
//! ```json
//! {"schema": 1,
//!  "backbone": {"type": "modernbert", "prefix": "encoder.", "hidden_size": 1024, ...},
//!  "head": {"type": "laya", ...},
//!  "weights": {"format": "safetensors", "file": "model.safetensors"}}
//! ```
//!
//! A model whose manifest has no arch layer runs on ONNX Runtime only.

use std::collections::BTreeMap;
use std::path::Path;

use serde::Deserialize;

use crate::Error;

pub const SCHEMA: u32 = 1;

#[derive(Debug, Clone, Deserialize)]
pub struct Arch {
    pub schema: u32,
    pub backbone: Backbone,
    pub head: Head,
    pub weights: Weights,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "type", rename_all = "kebab-case")]
pub enum Backbone {
    Modernbert(ModernBertConfig),
}

/// Hugging Face `ModernBertConfig` (transformers 5), as the encoder's `config.json` gives it.
#[derive(Debug, Clone, Deserialize)]
pub struct ModernBertConfig {
    /// Name prefix of the encoder's tensors in the weights file, e.g. `encoder.` or `model.`.
    pub prefix: String,
    pub vocab_size: usize,
    pub hidden_size: usize,
    pub intermediate_size: usize,
    pub num_hidden_layers: usize,
    pub num_attention_heads: usize,
    pub norm_eps: f32,
    pub norm_bias: bool,
    pub attention_bias: bool,
    pub mlp_bias: bool,
    /// Only `gelu` (exact, erf) is supported.
    pub hidden_activation: String,
    /// Window of the sliding-attention layers: a token sees `local_attention / 2` on each side.
    pub local_attention: usize,
    /// `full_attention` or `sliding_attention`, one per layer.
    pub layer_types: Vec<String>,
    /// RoPE base per layer type.
    pub rope_theta: BTreeMap<String, f32>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "type", rename_all = "kebab-case")]
pub enum Head {
    /// Laya's `DecisionModel` head (`laya-markers-v1`).
    Laya(LayaHead),
    /// `ModernBertForSequenceClassification` (`nli-pairs-v1`).
    SequenceClassification(SequenceClassificationHead),
    /// Von's `OptionMarkerScorer` (`von-option-marker-v1`).
    OptionMarker(OptionMarkerHead),
    /// GLiClass `GLiClassUniEncoder` (`gliclass-uni-v1`).
    GliclassUni(GliclassHead),
}

impl Head {
    pub fn name(&self) -> &'static str {
        match self {
            Head::Laya(_) => "laya",
            Head::SequenceClassification(_) => "sequence-classification",
            Head::OptionMarker(_) => "option-marker",
            Head::GliclassUni(_) => "gliclass-uni",
        }
    }
}

/// `laya.common.DecisionModel`: a type embedding, `layers` post-encoder `nn.TransformerEncoder`
/// layers (pre-norm, ReLU), a marker scorer and the act head.
#[derive(Debug, Clone, Deserialize)]
pub struct LayaHead {
    /// `type_emb.weight`, [3, hidden].
    pub type_embeddings: String,
    /// `head.layers.`: `nn.TransformerEncoderLayer` tensors (`self_attn.in_proj_weight`, ...).
    pub layers_prefix: String,
    pub layers: usize,
    pub num_heads: usize,
    /// Must be `relu` (PyTorch's default) with `norm_first`.
    pub activation: String,
    pub norm_first: bool,
    pub layer_norm_eps: f32,
    /// `scorer.`: LayerNorm (`0`), Linear (`1`), GELU, Linear (`3`).
    pub scorer_prefix: String,
    pub scorer_norm_eps: f32,
    /// `act_head.`: Linear (`0`), GELU, Linear (`2`), over [CLS state, 4 features].
    pub act_prefix: String,
}

/// `ModernBertForSequenceClassification`: pooling, `head` (dense, activation, norm), `classifier`.
#[derive(Debug, Clone, Deserialize)]
pub struct SequenceClassificationHead {
    /// `head.`: `dense`, `norm`.
    pub head_prefix: String,
    /// `classifier.`
    pub classifier_prefix: String,
    /// `mean` or `cls`.
    pub pooling: String,
    /// Only `gelu`.
    pub activation: String,
    pub classifier_bias: bool,
    pub num_labels: usize,
}

/// Von's `OptionMarkerScorer`: LayerNorm, Linear, GELU, LayerNorm, Linear, at every marker.
#[derive(Debug, Clone, Deserialize)]
pub struct OptionMarkerHead {
    /// `scorer.`: `input_norm`, `dense`, `norm`, `out_proj`.
    pub prefix: String,
    pub norm_eps: f32,
}

/// GLiClass `GLiClassUniEncoder`: segment embeddings added before the encoder, label-span mean
/// pooling, two `FeaturesProjector`s and an `MLPScorer`.
#[derive(Debug, Clone, Deserialize)]
pub struct GliclassHead {
    /// `<<SEP>>`: segment 1 starts at its first occurrence (all 1 when there is none).
    pub text_token_id: i32,
    /// `model.segment_embeddings.weight`, [3, hidden].
    pub segment_embeddings: String,
    /// The text embedding: only `first` (the hidden state at 0).
    pub pooling: String,
    /// `model.text_projector.` and `model.classes_projector.`: `linear_1`, activation, `linear_2`.
    pub text_projector: String,
    pub classes_projector: String,
    /// Only `gelu`.
    pub projector_activation: String,
    /// Only `mlp`: `mlp.0`, ReLU, `mlp.2`, ReLU, `mlp.4` over [text, label].
    pub scorer_type: String,
    /// `model.scorer.`
    pub scorer_prefix: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct Weights {
    pub format: WeightsFormat,
    /// The upstream file name (a development directory holds it under this name).
    pub file: String,
    /// For `torch-zip`: where each tensor's bytes are (the pickle is never read at run time).
    #[serde(default)]
    pub tensors: BTreeMap<String, TensorAt>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum WeightsFormat {
    Safetensors,
    TorchZip,
}

/// One tensor's raw little-endian bytes in the weights file.
#[derive(Debug, Clone, Deserialize)]
pub struct TensorAt {
    /// `F32`, `F16` or `BF16`.
    pub dtype: String,
    pub shape: Vec<usize>,
    /// Absolute byte offset in the file.
    pub offset: u64,
}

impl Arch {
    pub fn read(path: &Path) -> Result<Arch, Error> {
        let text = std::fs::read_to_string(path)
            .map_err(|e| Error::Model(format!("{}: {e}", path.display())))?;
        Arch::parse(&text).map_err(|e| Error::Model(format!("{}: {e}", path.display())))
    }

    pub fn parse(text: &str) -> Result<Arch, String> {
        let arch: Arch = serde_json::from_str(text).map_err(|e| e.to_string())?;
        if arch.schema != SCHEMA {
            return Err(format!(
                "arch schema {} (this build reads {SCHEMA}); upgrade ollaya",
                arch.schema
            ));
        }
        let Backbone::Modernbert(c) = &arch.backbone;
        if c.hidden_activation != "gelu" {
            return Err(format!(
                "unsupported hidden_activation {:?}",
                c.hidden_activation
            ));
        }
        if c.layer_types.len() != c.num_hidden_layers
            || c.num_attention_heads == 0
            || c.hidden_size % c.num_attention_heads != 0
        {
            return Err("inconsistent backbone".into());
        }
        if let Some(t) = c
            .layer_types
            .iter()
            .find(|t| !c.rope_theta.contains_key(*t))
        {
            return Err(format!("no rope_theta for {t}"));
        }
        Ok(arch)
    }
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::*;

    fn arch(activation: &str, schema: u32) -> String {
        json!({
            "schema": schema,
            "backbone": {
                "type": "modernbert", "prefix": "model.", "vocab_size": 10, "hidden_size": 8,
                "intermediate_size": 12, "num_hidden_layers": 2, "num_attention_heads": 2,
                "norm_eps": 1e-5, "norm_bias": false, "attention_bias": false, "mlp_bias": false,
                "hidden_activation": activation, "local_attention": 4,
                "layer_types": ["full_attention", "sliding_attention"],
                "rope_theta": {"full_attention": 160000.0, "sliding_attention": 10000.0}
            },
            "head": {"type": "option-marker", "prefix": "scorer.", "norm_eps": 1e-5},
            "weights": {"format": "torch-zip", "file": "option_marker.pt",
                        "tensors": {"scorer.dense.weight": {"dtype": "F32", "shape": [4, 8], "offset": 64}}}
        })
        .to_string()
    }

    #[test]
    fn reads_what_arch_py_writes() {
        let a = Arch::parse(&arch("gelu", SCHEMA)).unwrap();
        assert_eq!(a.head.name(), "option-marker");
        assert_eq!(a.weights.format, WeightsFormat::TorchZip);
        assert_eq!(a.weights.tensors["scorer.dense.weight"].offset, 64);
    }

    #[test]
    fn refuses_what_it_cannot_run() {
        assert!(
            Arch::parse(&arch("gelu", SCHEMA + 1))
                .unwrap_err()
                .contains("upgrade")
        );
        assert!(
            Arch::parse(&arch("silu", SCHEMA))
                .unwrap_err()
                .contains("silu")
        );
    }
}
